sdk/core/azure-core/src/io/body_stream.cpp

```diff
@@ -190,3 +190,32 @@ size_t FileBodyStream::OnRead(uint8_t* buffer, size_t count, Azure::Core::Contex
 void FileBodyStream::Rewind() { m_randomAccessFileBodyStream->Rewind(); }
 
 int64_t FileBodyStream::Length() const { return m_randomAccessFileBodyStream->Length(); }
+
+ProgressBodyStream::ProgressBodyStream(
+    BodyStream& bodyStream,
+    std::function<void(int64_t bytesTransferred)> callback)
+    : m_bodyStream(bodyStream), m_bytesTransferred(0), m_callback(std::move(callback))
+{
+  m_callback(m_bytesTransferred);
+}
+
+void ProgressBodyStream::Rewind()
+{
+  m_bodyStream.Rewind();
```

Comment:
Depending on the type of bodystream is passed-in by the user, this will fail a precondition. Do we need to support this scenario? And if so, should we let the user know (maybe even just as a doc comment) what types of streams are supported?

https://github.com/Azure/azure-sdk-for-cpp/blob/7d7d0361367911700ec184db496544596ebc0e63/sdk/core/azure-core/inc/azure/core/io/body_stream.hpp#L64-L71

Reply:
All derived types have Rewind. 
If future types do not have it it will act just as if the call would be done directly on the original stream, so we're not changing behavior. 
Since BaseStream is abstract you cannot instantiate it , only derived classes.